more_itertools/more.py

```diff
@@ -949,7 +1105,35 @@ def walk(node, level):
     yield from walk(iterable, 0)
 
 
-def side_effect(func, iterable, chunk_size=None, before=None, after=None):
+@overload
+def side_effect(
+    func: Callable[[_T], object],
+    iterable: Iterable[_T],
+    chunk_size: None = ...,
+    before: Optional[Callable[[], object]] = ...,
+    after: Optional[Callable[[], object]] = ...,
+) -> Iterator[_T]:
+    ...
+
+
+@overload
+def side_effect(
+    func: Callable[[List[_T]], object],
```

Comment:
Is `List` unnecessarily constraining here?  Shouldn't any function which can take an `Iterable[_T]` work?  (personally, I very rarely write any function with an argument type of `List` - it's usually either `Iterable` or `Sequence`, depending on what I need to do with it)

Reply:
`Callable[[List[_T]], object]` is actually less restrictive than `Callable[[Iterable[_T]], object]`, because any function that can take an iterable as an argument can take a list as an argument. Functions with an argument type of  `Iterable[_T]`, `Collection[_T]`, or any other supertype of `List[_T]` will be accepted.

```python
def f(a: Callable[[List[str]], object]):
    ...

def g(a: Callable[[Sequence[str]], object]):
    ...

def u(x: List[str]):
    ...

def v(x: Sequence[str]):
    ...

f(u)
f(v)
g(u)  #  Argument 1 to "g" has incompatible type "Callable[[List[str]], Any]"; expected "Callable[[Sequence[str]], object]"  [arg-type]
g(v)
```

In this example, both `u` and `v` are accepted as arguments of `f`, but only `v` is a valid argument for `g`.

Reply:
Thanks!